### PR TITLE
Obtaining child AST text span

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
Adds utilities to check the position of AST nodes in the text relative to their parents.
Also, fixed a number of byte/codepoint indexing mismatches and added some tests for this (though still coverage could be improved).

`impl std::ops::Index<Span> for str ` still needs more refactoring, will finish this tomorrow.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/main/docs/style-guide/rust.md).
- [x] All code has been tested where possible.
- [ ] All code has been profiled where possible.

### CODE REVIEW

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 1]
```diff
@@ -303,6 +310,32 @@ impl Ast {
...
+            } else if !found_child {
+                position += token.len()
+            }
+        });
+        if found_child {
+            Ok(Index::new(position))
+        } else {
+            Err(NoSuchChild.into())
+        }
+    }
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:331]: `Index` term in this context is really misleading for me. It's not natural that this means "index of character". Usually we use rather "offset" term.

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 2]
```diff
@@ -852,8 +890,8 @@ impl TokenConsumer for IdMapBuilder {
     fn feed(&mut self, token:Token) {
         match token {
             Token::Off(val) => self.offset += val,
-            Token::Chr( _ ) => self.offset += 1,
-            Token::Str(val) => self.offset += val.len(),
+            Token::Chr(val) => self.offset += val.len_utf8(),
+            Token::Str(val) => self.offset += val.chars().count(),
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:894]: Once we add the number of bytes, and then chars count?

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 3]
```diff
@@ -896,8 +934,8 @@ impl TokenConsumer for LengthBuilder {
     fn feed(&mut self, token:Token) {
         match token {
             Token::Off(val) => self.length += val,
-            Token::Chr( _ ) => self.length += 1,
-            Token::Str(val) => self.length += val.len(),
+            Token::Chr(val) => self.length += val.len_utf8(),
+            Token::Str(val) => self.length += val.chars().count(),
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:938]: Ditto

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 4]
```diff
@@ -911,6 +949,19 @@ impl<T:HasTokens> HasLength for T {
     }
 }
 
+impl HasLength for Token<'_> {
+    fn len(&self) -> usize {
+        match self {
+            Token::Off(val) => *val,
+            Token::Chr(val) => val.len_utf8(),
+            Token::Str(val) => val.chars().count(),
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:957]: Ditto

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 5]
```diff
@@ -1004,8 +1055,8 @@ where F:FnMut(Index,&Ast) {
     fn feed(&mut self, token:Token) {
         match token {
             Token::Off(val) => self.index += val,
-            Token::Chr( _ ) => self.index += 1,
-            Token::Str(val) => self.index += val.len(),
+            Token::Chr(val) => self.index += val.len_utf8(),
+            Token::Str(val) => self.index += val.chars().count(),
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:1059]: Ditto

File: src/rust/lib/data/src/text.rs [hunk 6]
```diff
@@ -113,7 +113,7 @@ impl Display for Size {
 
 impl From<&str> for Size {
     fn from(text:&str) -> Self {
-        Size::new(text.len())
+        Size::new(text.chars().count())
     }
 }
```
Comment (farmaazon) [inline src/rust/lib/data/src/text.rs:118]: This is not a today change, but I have distaste for this conversion. What does it mean that string is _converted_ to its length? Those are completely different things.

File: src/rust/ide/lib/ast/impl/src/lib.rs (hunk 1 above)
Comment (mwu-tow) [reply src/rust/ide/lib/ast/impl/src/lib.rs:331]: -> child_offset

File: src/rust/ide/lib/ast/impl/src/lib.rs (hunk 2 above)
Comment (mwu-tow) [reply src/rust/ide/lib/ast/impl/src/lib.rs:894]: Stupid mistake. One was wrong and one was correct but I just mindlessly flipped both. Fixed.

File: src/rust/lib/data/src/text.rs (hunk 6 above)
Comment (mwu-tow) [reply src/rust/lib/data/src/text.rs:118]: Agreed as discussed today. The impl has been removed in favor of `from_str` constructor.

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 7]
```diff
@@ -311,8 +311,8 @@ impl Ast {
         self.iter_recursive().find(|ast| ast.repr() == repr)
     }
 
-    /// Get the index (relative to self) for given child node.
-    pub fn index_of_child(&self, child:&Ast) -> FallibleResult<Index> {
+    /// Get the offset relative to self for a given child node.
```
Comment (farmaazon) [inline src/rust/ide/lib/ast/impl/src/lib.rs:314]: I would still extend this docs. To be more clear that this is an offest of "begins of text representation" in number of codepoints (although the last can be deduced from the returned type).

File: src/rust/lib/data/src/text.rs [hunk 8]
```diff
@@ -152,6 +167,17 @@ impl Span {
         Span {index:Index::new(0), size}
     }
 
+    /// Get the index of the last character in the span.
+    ///
+    /// If the span is empty or starts at 0, returns `None`.
```
Comment (farmaazon) [inline src/rust/lib/data/src/text.rs:172]:
```suggestion
    /// If the span is empty or ends at 0, returns `None`.
```

File: src/rust/lib/data/src/text.rs [hunk 9]
```diff
@@ -152,6 +167,17 @@ impl Span {
...
+    pub fn last(&self) -> Option<Index> {
+        if self.is_empty() {
+            None
+        } else {
+            self.end().checked_sub(Size::new(1))
+        }
+    }
```
Comment (farmaazon) [inline src/rust/lib/data/src/text.rs:179]: Technically, it is impossible now to have span which is not empty, but it end is at 0. But maybe we should leave it as it is, and fix when we will handle panics correctly.

File: src/rust/lib/data/src/text.rs (hunk 8 above)
Comment (mwu-tow) [reply src/rust/lib/data/src/text.rs:172]: I've removed that part, as you mentioned that this situation is impossible.

File: src/rust/lib/data/src/text.rs (hunk 9 above)
Comment (mwu-tow) [reply src/rust/lib/data/src/text.rs:179]: True.

File: src/rust/ide/lib/ast/impl/src/lib.rs (hunk 7 above)
Comment (mwu-tow) [reply src/rust/ide/lib/ast/impl/src/lib.rs:314]: I've extended somehow, though didn't feel like mentioning codepoints. The type expresses it, we should not need to repeat it on every usage.

File: src/rust/ide/lib/ast/impl/src/lib.rs [hunk 10]
```diff
@@ -767,6 +803,11 @@ pub trait TokenConsumer {
     fn feed(&mut self, val:Token);
 }
 
+impl<F:FnMut(Token)> TokenConsumer for F {
+    fn feed(&mut self, val: Token) {
```
Comment (wdanilo) [inline src/rust/ide/lib/ast/impl/src/lib.rs:807]: rm space after :

**Pull request merged**